File: src/main/java/org/taktik/icure/entities/Message.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.entities;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.google.common.base.Objects;
import org.taktik.icure.entities.base.StoredICureDocument;
import org.taktik.icure.entities.embed.Delegation;
import org.taktik.icure.entities.utils.MergeUtil;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonIgnoreProperties(ignoreUnknown = true)
public class Message extends StoredICureDocument implements Serializable {

	public final static int STATUS_LABO_RESULT 				= 1 << 0;
	public final static int STATUS_UNREAD 					= 1 << 1;
	public final static int STATUS_IMPORTANT 				= 1 << 2;
	public final static int STATUS_ENCRYPTED 				= 1 << 3;
	public final static int STATUS_HAS_ANNEX 				= 1 << 4;
	public final static int STATUS_HAS_FREE_INFORMATION 	= 1 << 5;
	public final static int STATUS_EFACT 					= 1 << 6;

	public final static int STATUS_SENT  	     			= 1 << 7;
	public final static int STATUS_SUBMITTED 				= 1 << 8; //tack
	public final static int STATUS_RECEIVED 				= 1 << 9; //tack
	public final static int STATUS_ACCEPTED_FOR_TREATMENT 	= 1 << 10; //931000
	public final static int STATUS_ACCEPTED 				= 1 << 11; //920098
	public final static int STATUS_REJECTED 				= 1 << 12; //920999

	public final static int STATUS_TACK 					= 1 << 13;
	public final static int STATUS_MASKED 					= 1 << 14;

	public final static int STATUS_SUCCESS 					= 1 << 15; //920900 920098
	public final static int STATUS_WARNING 					= 1 << 16; //920900
	public final static int STATUS_ERROR 					= 1 << 17; //920099

	public final static int STATUS_ANALYZED 				= 1 << 18;
	public final static int STATUS_DELETED_ON_SERVER 		= 1 << 19;
	public final static int STATUS_SHOULD_BE_DELETED_ON_SERVER 	= 1 << 20;

	public final static int STATUS_ARCHIVED					= 1 << 21;

	private String fromAddress;
	private String fromHealthcarePartyId;
	private String formId;
	private Integer status;
	private String recipientsType;

	private Set<String> recipients = new HashSet<>(); //The id of the hcp whose the message is addressed to
	private Set<String> toAddresses = new HashSet<>(); //The address of the recipient of the message. Format is of an email address with extra domains defined for mycarenet and ehealth: (efact.mycarenet.be/eattest.mycarenet.be/chapter4.mycarenet.be/ehbox.ehealth.fgov.be)

	private Long received;
	private Long sent;

	private Map<String, String> metas = new HashMap<>();

	/*
		CHAP4:IN:   ${Mycarenet message ref}
		CHAP4:OUT:  ${Mycarenet message ref}
		EFACT:BATCH:${iCure batch ref}
		EFACT:IN:   ${Mycarenet message ref}
		EFACT:OUT:  ${Mycarenet message ref}
		GMD:IN:     ${Mycarenet message ref}
		INBOX:      ${Ehealth box message ref}
		SENTBOX:    ${Ehealth box message ref}
		BININBOX:   ${Ehealth box message ref}
		BINSENTBOX: ${Ehealth box message ref}
		REPORT:IN:  ${iCure ref}
		REPORT:OUT: ${iCure ref}
	 */

	private String transportGuid; //Each message should have a transportGuid: see above for formats
	private String remark;
	private String conversationGuid;

	private String subject;

	private List<String> invoiceIds = new ArrayList<>();
	private String parentId; //ID of parent in a message conversation
	private String externalRef;

	private Set<String> unassignedResults; //refs
	private Map<String, String> assignedResults; //ContactId -> ref

	private Map<String,String> senderReferences;

	public Message solveConflictWith(Message other) {
		super.solveConflictsWith(other);

		this.fromAddress = this.fromAddress == null ? other.fromAddress : this.fromAddress;
		this.fromHealthcarePartyId = this.fromHealthcarePartyId == null ? other.fromHealthcarePartyId : this.fromHealthcarePartyId ;
		this.formId = this.formId == null ? other.formId : this.formId;
		this.recipients.addAll(other.recipients);
		this.toAddresses.addAll(other.toAddresses);

		this.received = other.received==null?this.received:this.received==null?other.received:Long.valueOf(Math.min(this.received,other.received));
		this.sent = other.sent==null?this.sent:this.sent==null?other.sent:Long.valueOf(Math.min(this.sent,other.sent));

		this.remark = this.remark == null ? other.remark : this.remark;
		this.transportGuid = this.transportGuid == null ? other.transportGuid : this.transportGuid ;
		this.conversationGuid = this.conversationGuid == null ? other.conversationGuid : this.conversationGuid;

		this.subject = this.subject == null ? other.subject : this.subject;
		this.parentId = this.parentId == null ? other.parentId : this.parentId ;
		this.externalRef = this.externalRef == null ? other.externalRef : this.externalRef;

		this.invoiceIds = MergeUtil.mergeListsDistinct(this.invoiceIds,other.invoiceIds, Objects::equal, (a, b)->a);
		other.metas.forEach((k,v)->this.metas.putIfAbsent(k,v));

		return this;
	}

	public String getFormId() {
        return formId;
    }

    public void setFormId(String formId) {
        this.formId = formId;
    }

    public Integer getStatus() {
        return status;
    }

    public void setStatus(Integer status) {
        this.status = status;
    }

    public String getFromAddress() {
        return fromAddress;
    }

    public void setFromAddress(String fromAddress) {
        this.fromAddress = fromAddress;
    }

    public Set<String> getToAddresses() {
        return toAddresses;
    }

    public void setToAddresses(Set<String> toAddresses) {
        this.toAddresses = toAddresses;
    }

    public String getFromHealthcarePartyId() {
        return fromHealthcarePartyId;
    }

    public void setFromHealthcarePartyId(String fromHealthcarePartyId) {
        this.fromHealthcarePartyId = fromHealthcarePartyId;
    }

    public Long getReceived() {
        return received;
    }

    public void setReceived(Long received) {
        this.received = received;
    }

    public Long getSent() {
        return sent;
    }

    public void setSent(Long sent) {
        this.sent = sent;
    }

    public Map<String, String> getMetas() {
        return metas;
    }

    public void setMetas(Map<String, String> metas) {
        this.metas = metas;
    }

    public String getTransportGuid() {
        return transportGuid;
    }

    public void setTransportGuid(String transportGuid) {
        this.transportGuid = transportGuid;
    }

    public String getRemark() {
        return remark;
    }

    public void setRemark(String remark) {
        this.remark = remark;
    }

    public String getConversationGuid() {
        return conversationGuid;
    }

    public void setConversationGuid(String conversationGuid) {
        this.conversationGuid = conversationGuid;
    }

    public String getSubject() {
        return subject;
    }

    public void setSubject(String subject) {
        this.subject = subject;
    }

    @JsonIgnore
    public Set<String> getSecretContactKeys() {
        return super.getSecretForeignKeys();
    }

    @JsonIgnore
    public void setSecretContactKeys(Set<String> secretContactKeys) {
        super.setSecretForeignKeys(secretContactKeys);
    }

    @JsonIgnore
    public Map<String, Set<Delegation>> getCryptedContactIds() {
        return super.getCryptedForeignKeys();
    }

    @JsonIgnore
    public void setCryptedContactIds(Map<String, Set<Delegation>> cryptedContactIds) {
        super.setCryptedForeignKeys(cryptedContactIds);
    }

	public Set<String> getRecipients() {
		return recipients;
	}

	public void setRecipients(Set<String> recipients) {
		this.recipients = recipients;
	}

	public List<String> getInvoiceIds() {
		return invoiceIds;
	}

	public void setInvoiceIds(List<String> invoiceIds) {
		this.invoiceIds = invoiceIds;
	}

	public String getParentId() {
		return parentId;
	}

	public void setParentId(String parentId) {
		this.parentId = parentId;
	}

	public void setRecipientsType(String recipientsType) {
		this.recipientsType = recipientsType;
	}

	public String getRecipientsType() {
		return recipientsType;
	}

	public void setExternalRef(String externalRef) {
		this.externalRef = externalRef;
	}

	public String getExternalRef() {
		return externalRef;
	}

	private String encryptedSelf;
	@Override
	public String getEncryptedSelf() {
		return encryptedSelf;
	}

	@Override
	public void setEncryptedSelf(String encryptedSelf) {
		this.encryptedSelf = encryptedSelf;
	}

	public Map<String, String> getSenderReferences() {
		return senderReferences;
	}

	public void setSenderReferences(Map<String, String> senderReferences) {
		this.senderReferences = senderReferences;
	}

	public Set<String> getUnassignedResults() {
		return unassignedResults;
	}

	public void setUnassignedResults(Set<String> unassignedResults) {
		this.unassignedResults = unassignedResults;
	}

	public Map<String, String> getAssignedResults() {
		return assignedResults;
	}

	public void setAssignedResults(Map<String, String> assignedResults) {
		this.assignedResults = assignedResults;
	}
}

File: src/main/kotlin/org/taktik/icure/entities/Message.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */
package org.taktik.icure.entities

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonProperty
import com.github.pozo.KotlinBuilder
import org.taktik.couchdb.entity.Attachment
import org.taktik.icure.entities.base.CodeStub
import org.taktik.icure.entities.base.Encryptable
import org.taktik.icure.entities.base.StoredICureDocument
import org.taktik.icure.entities.embed.Delegation
import org.taktik.icure.entities.embed.MessageReadStatus
import org.taktik.icure.entities.embed.RevisionInfo
import org.taktik.icure.utils.DynamicInitializer
import org.taktik.icure.utils.invoke
import org.taktik.icure.validation.AutoFix
import org.taktik.icure.validation.NotNull
import org.taktik.icure.validation.ValidCode

@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonIgnoreProperties(ignoreUnknown = true)
@KotlinBuilder
data class Message(
        @JsonProperty("_id") override val id: String,
        @JsonProperty("_rev") override val rev: String? = null,
        @field:NotNull(autoFix = AutoFix.NOW) override val created: Long? = null,
        @field:NotNull(autoFix = AutoFix.NOW) override val modified: Long? = null,
        @field:NotNull(autoFix = AutoFix.CURRENTUSERID) override val author: String? = null,
        @field:NotNull(autoFix = AutoFix.CURRENTHCPID) override val responsible: String? = null,
        override val medicalLocationId: String? = null,
        @field:ValidCode(autoFix = AutoFix.NORMALIZECODE) override val tags: Set<CodeStub> = emptySet(),
        @field:ValidCode(autoFix = AutoFix.NORMALIZECODE) override val codes: Set<CodeStub> = emptySet(),
        override val endOfLife: Long? = null,
        @JsonProperty("deleted") override val deletionDate: Long? = null,

        val fromAddress: String? = null,
        val fromHealthcarePartyId: String? = null,
        @Deprecated("Never used") val formId: String? = null,
        val status: Int? = null,
        val recipientsType: String? = null,
        val recipients: Set<String> = emptySet(), //The id of the hcp whose the message is addressed to
        val toAddresses: Set<String> = emptySet(), //The address of the recipient of the message. Format is of an email address with extra domains defined for mycarenet and ehealth: (efact.mycarenet.be/eattest.mycarenet.be/chapter4.mycarenet.be/ehbox.ehealth.fgov.be)
        val received: Long? = null,
        val sent: Long? = null,
        val metas: Map<String, String> = emptyMap(),
        val readStatus: Map<String, MessageReadStatus> = emptyMap(),
        /*
            CHAP4:IN:   ${Mycarenet message ref}
            CHAP4:OUT:  ${Mycarenet message ref}
            EFACT:BATCH:${Mycarenet message ref}
            EFACT:IN:   ${Mycarenet message ref}
            EFACT:OUT:  ${Mycarenet message ref}
            GMD:IN:     ${Mycarenet message ref}
            INBOX:      ${Ehealth box message ref}
            SENTBOX:    ${Ehealth box message ref}
            BININBOX:   ${Ehealth box message ref}
            BINSENTBOX: ${Ehealth box message ref}
            REPORT:IN:  ${iCure ref}
            REPORT:OUT: ${iCure ref}
         */
        val transportGuid: String? = null, //Each message should have a transportGuid: see above for formats
        val remark: String? = null,
        val conversationGuid: String? = null,
        val subject: String? = null,
        val invoiceIds: Set<String> = emptySet(),
        val parentId: String? = null, //ID of parent in a message conversation
        val externalRef: String? = null,
        val unassignedResults: Set<String> = emptySet(), //refs
        val assignedResults: Map<String, String> = emptyMap(), //ContactId -> ref
        val senderReferences: Map<String, String> = emptyMap(),

        override val secretForeignKeys: Set<String> = emptySet(),
        override val cryptedForeignKeys: Map<String, Set<Delegation>> = emptyMap(),
        override val delegations: Map<String, Set<Delegation>> = emptyMap(),
        override val encryptionKeys: Map<String, Set<Delegation>> = emptyMap(),
        override val encryptedSelf: String? = null,
        @JsonProperty("_attachments") override val attachments: Map<String, Attachment>? = emptyMap(),
        @JsonProperty("_revs_info") override val revisionsInfo: List<RevisionInfo>? = emptyList(),
        @JsonProperty("_conflicts") override val conflicts: List<String>? = emptyList(),
        @JsonProperty("rev_history") override val revHistory: Map<String, String>? = emptyMap()

) : StoredICureDocument, Encryptable {
    companion object : DynamicInitializer<Message> {
        const val STATUS_LABO_RESULT = 1 shl 0
        const val STATUS_UNREAD = 1 shl 1
        const val STATUS_IMPORTANT = 1 shl 2
        const val STATUS_ENCRYPTED = 1 shl 3
        const val STATUS_HAS_ANNEX = 1 shl 4
        const val STATUS_HAS_FREE_INFORMATION = 1 shl 5
        const val STATUS_EFACT = 1 shl 6
        const val STATUS_SENT = 1 shl 7
        const val STATUS_SUBMITTED = 1 shl 8 //tack
        const val STATUS_RECEIVED = 1 shl 9 //tack
        const val STATUS_ACCEPTED_FOR_TREATMENT = 1 shl 10 //931000
        const val STATUS_ACCEPTED = 1 shl 11 //920098 920900 920099
        const val STATUS_REJECTED = 1 shl 12 //920999
        const val STATUS_TACK = 1 shl 13
        const val STATUS_MASKED = 1 shl 14
        const val STATUS_FULL_SUCCESS = 1 shl 15 //920900 920098
        const val STATUS_PARTIAL_SUCCESS = 1 shl 16 //920900
        const val STATUS_FULL_ERROR = 1 shl 17 //920099 920999
        const val STATUS_ANALYZED = 1 shl 18
        const val STATUS_DELETED_ON_SERVER = 1 shl 19
        const val STATUS_SHOULD_BE_DELETED_ON_SERVER = 1 shl 20
        const val STATUS_ARCHIVED = 1 shl 21
        const val STATUS_ERRORS_IN_PRELIMINARY_CONTROL = 1 shl 22
        const val STATUS_DRAFT = 1 shl 23
        const val STATUS_SCANNED = 1 shl 24
        const val STATUS_IMPORTED = 1 shl 25
        const val STATUS_TREATED = 1 shl 26
    }

    fun merge(other: Message) = Message(args = this.solveConflictsWith(other))
    fun solveConflictsWith(other: Message) = super<StoredICureDocument>.solveConflictsWith(other) + super<Encryptable>.solveConflictsWith(other) + mapOf(
            "fromAddress" to (this.fromAddress ?: other.fromAddress),
            "fromHealthcarePartyId" to (this.fromHealthcarePartyId ?: other.fromHealthcarePartyId),
            "formId" to (this.formId ?: other.formId),
            "status" to (this.status ?: other.status),
            "recipientsType" to (this.recipientsType ?: other.recipientsType),
            "recipients" to (other.recipients + this.recipients),
            "toAddresses" to (other.toAddresses + this.toAddresses),
            "received" to (this.received ?: other.received),
            "sent" to (this.sent ?: other.sent),
            "metas" to (other.metas + this.metas),
            "readStatus" to (this.readStatus),
            "transportGuid" to (this.transportGuid ?: other.transportGuid),
            "remark" to (this.remark ?: other.remark),
            "conversationGuid" to (this.conversationGuid ?: other.conversationGuid),
            "subject" to (this.subject ?: other.subject),
            "invoiceIds" to (other.invoiceIds + this.invoiceIds),
            "parentId" to (this.parentId ?: other.parentId),
            "externalRef" to (this.externalRef ?: other.externalRef),
            "unassignedResults" to (other.unassignedResults + this.unassignedResults),
            "assignedResults" to (other.assignedResults + this.assignedResults),
            "senderReferences" to (other.senderReferences + this.senderReferences)
    )

    override fun withIdRev(id: String?, rev: String) = if (id != null) this.copy(id = id, rev = rev) else this.copy(rev = rev)
    override fun withDeletionDate(deletionDate: Long?) = this.copy(deletionDate = deletionDate)
    override fun withTimestamps(created: Long?, modified: Long?) =
            when {
                created != null && modified != null -> this.copy(created = created, modified = modified)
                created != null -> this.copy(created = created)
                modified != null -> this.copy(modified = modified)
                else -> this
            }

}
